clang/runtime/dpct-rt/include/lapack_utils.hpp.inc

```diff
@@ -133,6 +132,260 @@ inline int sygvd(sycl::queue &queue, std::int64_t itype, oneapi::mkl::job jobz,
+  } catch (oneapi::mkl::lapack::exception const& e) {
+    std::cerr << "Unexpected exception caught during call to LAPACK API: hegvd"
+              << std::endl
+              << "reason: " << e.what() << std::endl
+              << "info: " << e.info() << std::endl;
+    int info_val = static_cast<int>(e.info());
+    queue.memcpy(info, &info_val, sizeof(int)).wait();
+    return 1;
+  } catch (sycl::exception const& e) {
+    std::cerr << "Caught synchronous SYCL exception:" << std::endl
+              << "reason: " << e.what() << std::endl;
+    queue.memset(info, 0, sizeof(int)).wait();
+    return 1;
+  }
+  queue.memset(info, 0, sizeof(int));
+  return 0;
+#endif
+}
+// DPCT_LABEL_END
+// DPCT_LABEL_BEGIN|potrf_batch|dpct::lapack
+// DPCT_DEPENDENCY_BEGIN
+// Util|DataType|UsmRestricted
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Computes the Cholesky factorizations of a batch of symmetric (or Hermitian,
+/// for complex data) positive-definite matrices.
+/// \return Returns 0 if no synchronous exception, otherwise returns 1.
+/// \param [in] queue Device queue where calculations will be performed.
+/// \param [in] uplo Must be uplo::upper or uplo::lower.
+/// \param [in] n The order of the matrix A.
+/// \param [in,out] a Array of pointers to matrix A.
+/// \param [in] lda The leading dimension of matrix A.
+/// \param [out] info If lapack synchronous exception is caught, the value
+/// returned from info() method of the exception is set to \p info.
+/// \param [in] group_size The batch size.
+template <typename T>
+inline int potrf_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
+                       T *a[], int lda, int *info, int group_size) {
+#ifdef DPCT_USM_LEVEL_NONE
+  throw std::runtime_error("this API is unsupported when USM level is none");
+#else
+  using Ty = typename DataType<T>::T2;
+  struct matrix_info_t {
+    oneapi::mkl::uplo uplo_info;
+    std::int64_t n_info;
+    std::int64_t lda_info;
+    std::int64_t group_size_info;
+  };
+  matrix_info_t *matrix_info =
+      (matrix_info_t *)std::malloc(sizeof(matrix_info_t));
+  matrix_info->uplo_info = uplo;
+  matrix_info->n_info = n;
+  matrix_info->lda_info = lda;
+  matrix_info->group_size_info = group_size;
+  std::int64_t scratchpad_size = 0;
+  sycl::event e;
+  Ty *scratchpad = nullptr;
+  try {
+    scratchpad_size = oneapi::mkl::lapack::potrf_batch_scratchpad_size<Ty>(
+        queue, &(matrix_info->uplo_info), &(matrix_info->n_info),
+        &(matrix_info->lda_info), 1, &(matrix_info->group_size_info));
+    scratchpad = sycl::malloc_device<Ty>(scratchpad_size, queue);
+    e = oneapi::mkl::lapack::potrf_batch(
+        queue, &(matrix_info->uplo_info), &(matrix_info->n_info), (Ty **)a,
+        &(matrix_info->lda_info), 1, &(matrix_info->group_size_info),
+        scratchpad, scratchpad_size);
+  } catch (oneapi::mkl::lapack::exception const &e) {
+    std::cerr << "Unexpected exception caught during call to LAPACK API: "
```

Comment:
Same comment as above applies here and to other synchronous exceptions' traps.

Reply:
Please note, `wait_and_throw()` should be called on the event object in order to deliver any async exceptions risen during execution of the GPU kernel.

Reply:
> Same comment as above applies here and to other synchronous exceptions' traps.

But as for now, due to some issues in compiler, LAPACK does not throw asynchronous exception, only reports calculation status as synchronous exception, so code in this review should be fine.

Reply:
Below is a general template on how to implement oneMKL LAPACK wrapper catching both synchronous and asynchronous exceptions without imposing additional requirements to the user's queue object (like "queue should be constructed with async error handling supporting these types of exception: ..."), and avoiding code duplication by implementing single catch block for both types of possible exceptions (sync/async):
```
#include <CL/sycl.hpp>

// some LAPACK function submitting kernel throwing asynchronous exception
sycl::event lapack_func(sycl::queue &q)
{
    return q.submit([&](auto &h){
            h.host_task([=]{
                throw std::exception();
            });
        });
}

// helper function enhancing arbitrary user provided queue with async error handling ability
sycl::queue get_error_handling_queue(sycl::queue q, std::exception_ptr &exception) {
    auto error_handler = [&] (sycl::exception_list exceptions) {
            exception = *exceptions.begin();
        };

    sycl::device  dev { q.get_device() };
    sycl::context ctx { dev, error_handler };
    return sycl::queue { ctx, dev };
}

// some lapack wrapper concerned with handling both sync/async exceptions
int lapack_wrapper_missed_exception(sycl::queue &q)
{
    // lapack async exception, if thrown, will be stored in this object
    std::exception_ptr lapack_async_exception;
    // enhance user queue with error handler catching any async lapack exception
    sycl::queue que = get_error_handling_queue(q, lapack_async_exception);
    try {
        // sychronous exception, if any, will happen here, and caught by catch block below
        sycl:: event e = lapack_func(que);
        // asyncronous exception, if any, will happen here:
        e.wait_and_throw();
        // rethrow async exception, if any happened, so it will be caught by same catch block below
        if (lapack_async_exception) {
            std::rethrow_exception(lapack_async_exception);
        }
    } catch (const std::exception &e) {
        printf("Caught: %s\n", e.what());
        return 0; // exception was not missed, test passed
    }
    return 1; // exception missed, test failed
}

int main()
{
    sycl::queue q{ sycl::gpu_selector_v };

    if (lapack_wrapper_missed_exception(q)) {
        printf("FAILED\n");
    } else {
        printf("PASSED\n");
    }
}
```

Reply:
The way demonstrated here avoids changes to common `exception_handler`: it makes call-site specific handlers, and exception traps process only exceptions specific to any particular function.

Reply:
Yes, but it introduced another queue and make this API to be synchronous.
If we need process the asynchronous exceptions, I think changing `exception_handler` is better than changing queue & synchronization.

Reply:
Ok, sounds good.